### DISCUSSION TITLE
Fixed Exact Time Clock

### DISCUSSION THREAD

**Discussion by m010-cpu** [FelixKratz/SketchyBar#480] [Q&A] [answered]
Hi, I'm trying to make a clock item with exact time (hh:mm:ss) but the problem is that every number doesn't have the same width and it makes the label moving. Does someone have any idea how I can fix it? Thanks in advance ! 

Tried using a monospace font but still getting the problem.

```
#!/usr/bin/env zsh

sketchybar --set $NAME label="$(date '+%H:%M:%S')"
```

```
#!/usr/bin/env zsh

clock=(
  label.color=$BLUE
  label.align=center
  update_freq=1
  script="$PLUGIN_DIR/clock.sh"
)

sketchybar --add item clock right \
    --set clock "${clock[@]}" \
    --subscribe clock system_woke
```

**Comment by johnallen3d** [comment, marked as answer]
If I understand your question correctly, I believe you're looking for a [Monospaced font](https://en.wikipedia.org/wiki/Monospaced_font) (or "fixed width").

**Reply by m010-cpu** [reply]
I've already tried a mono font but still have the issue.
I'm finally using fixed width like you recommended.
Thanks for your help!

**Comment by dav1d-wright** [comment]
I ran into exactly this issue and spent some time diagnosing it further, so sharing findings in case they help others.

**Fonts tried (for HH:MM:SS display)**

- FiraCode Nerd Font Mono Bold — visually non-monospace; digit widths varied by 1pt per frame
- FiraCode Nerd Font Mono Regular — same issue
- JetBrainsMono NFM Regular — still 1pt variation (77-78pt bounding rect over time)
- **Hack Nerd Font Mono Regular — stable at 78pt across 40 samples at 2Hz. This was the only font that was truly monospace in practice.**

**The padding issue with `width=N`**

Once we pinned the font, we noticed that adding `width=N` to stabilize the bounding rect broke inter-pill spacing: gaps between pills disappeared entirely.

Root cause is in the sketchybar source. In [`src/bar_manager.c`](https://github.com/FelixKratz/SketchyBar/blob/master/src/bar_manager.c), the layout function is:

```c
total_length += item_length + (bar_item->has_const_width
                               ? 0
                               : bar_item->background.padding_left
                                 + bar_item->background.padding_right);
```

When `width=N` is set, `has_const_width = true` and the ternary returns `0` — padding contributes nothing to the item's bar footprint. The item takes exactly N pixels and its bounding rect is placed edge-to-edge with neighbours. Confirmed with a live test: setting `padding_left=30` on a fixed-width item moved the adjacent item's x-coordinate by 0pt.

The relevant code is also in [`src/bar_item.c`](https://github.com/FelixKratz/SketchyBar/blob/master/src/bar_item.c) in `bar_item_get_length()`.

**Spacer attempt**

We tried inserting an invisible spacer item (`label.drawing=off`, `icon.drawing=off`, `background.drawing=off`, `padding_left=1`, `padding_right=1`) between the fixed-width clock and the adjacent dynamic item to restore the missing gap. The spacer does allocate space in the bar layout (confirmed visually), but the `--query` bounding_rect reports 0 for it, making gap verification unreliable.

**Resolution**

Switch to Hack Nerd Font Mono Regular for the clock. With a truly monospace font the bounding rect is stable without needing `width=N`, so `padding_left`/`padding_right` work normally and inter-pill gaps are consistent. No spacer needed.
